_gsocblogs/2024/blog_ATLCompression_IshanDarji.md

```diff
@@ -0,0 +1,120 @@
+
+Given the above comparisons, a natural question may be, why ztd? `Brotli` at first glance compresses more efficiently than `zstd`, and since it is the best metric why not choose that? There were a variety of factors that went into this decision: for one, `zstd` performs as a very good all-rounder. It compresses slightly faster than `brotli` and decompresses much faster, sacrificing only 0.5% of the original filesize in compression ratio for this win. Additionally, it offers a fairly wide range of potential settings in the advanced API, including using dictionaries to train better compression and streaming compression for files where we gain the input in fragments at a time. This opens up the possibility of future exploration that I was not able to look at this summer. Finally, `zstd` is guaranteed to be stable. The format used by `zstd` is very stable and well documented, the project is fully open source, it is guaranteed to be supported in the future and sees frequent maintenance/updates, it is already pre-bundled in many popular Linux distributions, and its use in the Linux kernel and CERN ROOT framework are evidence of its ubiquitiousness. For more information about `ztd`'s stability, its [RFC document](https://datatracker.ietf.org/doc/html/rfc8878) provides additional information about how data is compressed, along with security and longevity considerations. The last point was particularly important for CERN, where these files are expected to be compressed and kept for long periods of time till they are decompressed and used at some point in the future, so stability is a must.
+
+The next step was to go through ATLCopyBSEvent.cpp and modernize it, then integrate `zstd`. I started this process by stripping back the file such that it could only perform the basic operations that it needed: reading, writing, and copying files. Much of the code was removed with the capability to read and write from collections because it was deemed non-essential. Then, starting with this bare-bones version, I ran the program on some existing files that were pre-compressed to various levels, ensuring that it would still be possible to read the files in their varying states and write them out. Once this was ensured, I moved on to modernizing the code. In its old state, there was lots of room for misuse of objects with lots of raw C pointers being passed around, as well as lots of cases of small errors such as passing the raw pointer obtained from std::unique_ptr<T>.get() to a function, rather than moving it. Fixing these issues made the code appear much cleaner, and also meant that we got rid of any potential memory-related issues that could've arisen if the system were used in a larger context. I also added clang-format, to automatically format the code for consistency with the rest of the Atlas codebase.
```

Comment:
```suggestion
The next step was to go through `ATLCopyBSEvent`and modernize it, then integrate `zstd`. I started this process by stripping back the file such that it could only perform the basic operations that it needed: reading, writing, and copying files. Much of the code was removed with the capability to read and write from collections because it was deemed non-essential. Then, starting with this bare-bones version, I ran the program on some existing files that were pre-compressed to various levels, ensuring that it would still be possible to read the files in their varying states and write them out. Once this was ensured, I moved on to modernizing the code. In its old state, there was lots of room for misuse of objects with lots of raw C pointers being passed around, as well as lots of cases of small errors such as passing the raw pointer obtained from `std::unique_ptr<T>.get()` to a function, rather than moving it. Fixing these issues made the code appear much cleaner, and also meant that we got rid of any potential memory-related issues that could've arisen if the system were used in a larger context. I also added `clang-format`, to automatically format the code for consistency with the rest of the ATLAS codebase.
```